code/game/objects/items/robot/robot_upgrades.dm

```diff
@@ -343,93 +402,79 @@
 
 		cyborg = R
 		icon_state = "selfrepair_off"
-		toggle_action = new /datum/action/item_action/toggle(src)
-		toggle_action.Grant(R)
-
-/obj/item/borg/upgrade/selfrepair/deactivate(mob/living/silicon/robot/R, user = usr)
-	. = ..()
-	if (.)
-		if(toggle_action)
-			QDEL_NULL(toggle_action)
-		cyborg = null
-		deactivate_sr()
-
-/obj/item/borg/upgrade/selfrepair/dropped()
-	..()
-	addtimer(CALLBACK(src, .proc/check_dropped), 1)
-
-/obj/item/borg/upgrade/selfrepair/proc/check_dropped()
-	if(loc != cyborg)
-		if(toggle_action)
-			QDEL_NULL(toggle_action)
-		cyborg = null
-		deactivate_sr()
+		action = new /datum/action/item_action(src)
+		action.name = name
+		action.Grant(R)
 
 /obj/item/borg/upgrade/selfrepair/ui_action_click()
-	on = !on
-	if(on)
+	if(COOLDOWN_FINISHED(src, recharging_repmod) && icon_state == "selfrepair_off")
+		icon_state = "selfrepair_on"
 		to_chat(cyborg, "<span class='notice'>You activate the self-repair module.</span>")
-		START_PROCESSING(SSobj, src)
+		playsound(cyborg.loc, activation_sound, 30)
+		update_icon()
+		process()
```

Comment:
Process probably shouldn't be called outside of subsystems triggering it (Rename the proc)

Reply:
Should this call be async, since its using sleep?

Reply:
what is the issue with manually starting and stopping `process()` exactly? 

Process starts when the ability is activated, and halted when the ability is ended or if the module somehow ends up ejected. 

```
/obj/item/borg/upgrade/selfrepair/proc/deactivate_sr()
	STOP_PROCESSING(SSobj, src)
```

Process is used because it conveniently calls for the module to step forward every tick that it's active. I can't see any benefit to making new, unique proc that I have to set a separate timer to run identically to process 🤔

Reply:
Moved everything unnecessary outside of process, but I don't actually understand what the problem with manually starting and stopping process() on the item is?

Reply:
process() doesn't stop or sstart processing, START_PROCESSING does.
process() should only be called from the subsystem, as it needs to account for delta time.